src/CADL.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -519,3 +525,71 @@ export function getInputType(
         return namespaceString;
     }
 }
+
+export function getUsages(
+    program: Program,
+    ops?: HttpOperation[]
+): { inputs: string[]; outputs: string[]; roundTrips: string[] } {
+    const result: {
+        inputs: string[];
+        outputs: string[];
+        roundTrips: string[];
+    } = { inputs: [], outputs: [], roundTrips: [] };
+    if (!ops) {
+        return result;
+    }
+
+    const operations: Operation[] = ops.map((op) => op.operation);
+    const usages = resolveUsages(operations);
+    let usagesMap: Map<string, UsageFlags> = new Map<string, UsageFlags>();
+    for (const type of usages.types) {
+        let typeName = "";
+        if ("name" in type) typeName = type.name ?? "";
+        if (type.kind === "Model") {
+            const effectiveType = getEffectiveModelType(program, type);
+            typeName = effectiveType.name;
+        }
+        let affectTypes: string[] = [];
+        if (typeName !== "") affectTypes.push(typeName);
+        if (type.kind === "Model" && type.templateArguments) {
+            for (const arg of type.templateArguments) {
+                if (arg.kind === "Model" && "name" in arg && arg.name !== "") {
+                    affectTypes.push(arg.name);
+                }
+            }
+        }
+
+        for (const name of affectTypes) {
+            let value = usagesMap.get(name);
+            if (!value) value = UsageFlags.None;
+            if (usages.isUsedAs(type, UsageFlags.Input))
+                value = value | UsageFlags.Input;
+            if (usages.isUsedAs(type, UsageFlags.Output))
+                value = value | UsageFlags.Output;
+            usagesMap.set(name, value);
+        }
+    }
+    /* handle resource operation. */
+    for (const op of ops) {
+        const resourceOperation = getResourceOperation(program, op.operation);
+        if (resourceOperation) {
+            if (!op.parameters.bodyParameter && op.parameters.bodyType) {
+                const resourceName = resourceOperation.resourceType.name;
```

Comment:
Will this be changed if we enable discriminator?

Reply:
No. it is not related with discriminator.